Add validation to file-backed blob urls from WebContent

pascoej opened the pull request:
#### 9febfdbe1756b5f4924f13105a2a9a6da9d1915a
<pre>
Add validation to file-backed blob urls from WebContent
<a href="https://bugs.webkit.org/show_bug.cgi?id=276193">https://bugs.webkit.org/show_bug.cgi?id=276193</a>
<a href="https://rdar.apple.com/80200949">rdar://80200949</a>

Reviewed by Alex Christensen.

This validates that file paths are valid for a given WebContent process when registering the blob

* LayoutTests/editing/pasteboard/copy-paste-attachment.html:
* LayoutTests/fast/attachment/attachment-dom.html:
* LayoutTests/fast/attachment/attachment-folder-icon-expected.html:
* LayoutTests/fast/attachment/attachment-icon-from-file-extension-expected.html:
* LayoutTests/fast/attachment/attachment-label-highlight.html:
* LayoutTests/fast/attachment/attachment-progress.html:
* LayoutTests/fast/attachment/attachment-select-on-click-inside-user-select-all.html:
* LayoutTests/fast/attachment/attachment-select-on-click.html:
* LayoutTests/fast/attachment/attachment-subtitle.html:
* LayoutTests/fast/attachment/attachment-title.html:
* LayoutTests/fast/attachment/attachment-type-attribute-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-dom.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension-expected.html:
* LayoutTests/fast/attachment/mac/attachment-keynote-icon-expected.html:
* LayoutTests/fast/attachment/mac/attachment-numbers-icon-expected.html:
* LayoutTests/fast/attachment/mac/attachment-pages-icon-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute-expected.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-convert-to-download-async-delegate.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-convert-to-download.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-blank-base-target-popup-not-allowed.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-blank-target-popup-not-allowed.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-blank-target.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-backslash.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-doublequote.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-slashes.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-unicode.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-no-extension.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::registerBlobPathForTesting):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::createFile):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::isFilePathAllowed):
(WebKit::NetworkConnectionToWebProcess::registerInternalFileBlobURL):
(WebKit::NetworkConnectionToWebProcess::registerInternalBlobURLOptionallyFileBacked):
(WebKit::NetworkConnectionToWebProcess::registerBlobPathForTesting):
(WebKit::NetworkConnectionToWebProcess::allowAccessToFile):
(WebKit::NetworkConnectionToWebProcess::allowAccessToFiles):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
(WebKit::NetworkConnectionToWebProcess::blobFileAccessEnforcementEnabled const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::allowFilesAccessFromWebProcess):
(WebKit::NetworkProcess::allowFileAccessFromWebProcess):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::grantAccessToCurrentPasteboardData):
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::grantAccessToCurrentData):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::maybeInitializeSandboxExtensionHandle):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadDataWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::reload):
(WebKit::WebPageProxy::executeEditCommand):
(WebKit::WebPageProxy::performDragOperation):
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::contextMenuItemSelected):
(WebKit::WebPageProxy::didChooseFilesForOpenPanel):
(WebKit::WebPageProxy::platformDidSelectItemFromActiveContextMenu):
(WebKit::WebPageProxy::willPerformPasteCommand):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURL):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURLs):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _handleDOMPasteRequestWithResult:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::willPerformPasteCommand):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::readSelectionFromPasteboard):
(WebKit::WebPageProxy::platformDidSelectItemFromActiveContextMenu):
(WebKit::WebPageProxy::willPerformPasteCommand):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::performDragWithLegacyFiles):
(WebKit::handleLegacyFilesPasteboard):
(WebKit::WebViewImpl::performDragOperation):
(WebKit::WebViewImpl::requestDOMPasteAccess):
(WebKit::WebViewImpl::handleDOMPasteRequestForCategoryWithResult):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::registerBlobPathForTesting):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::registerBlobPathForTesting):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(TestWebKitAPI::TEST(WebKit, RunOpenPanelNonLatin1)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UploadDirectory.mm:
(TEST(WebKit, UploadDirectory)):

Canonical link: <a href="https://commits.webkit.org/281966@main">https://commits.webkit.org/281966@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d8f4496e663199f465d869c3fb33309ec60bbea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14168 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12159 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63703 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48619 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12399 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/65587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/8438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38033 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53338 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/65587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10570 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11057 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54678 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56514 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10869 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60824 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5524 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10633 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/67316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5550 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53293 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/67316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13719 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4545 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82587 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36737 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14435 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37822 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38916 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37567 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -778,6 +778,20 @@ BeaconAPIEnabled:
     WebCore:
       default: false
 
+BlobFileAccessEnforceEnabled:
```
Enforcement

achristensen07 commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -37184,10 +37184,9 @@
 				FFEFAB2D183BCC6F00514549 /* SubscriptionObserver.cpp */,
-				FFEFAB2D183BCC6F00514547 /* SubscriptionObserver.h */,
-				FFEFAB2D183BCC6F00514548 /* SubscriptionObserver.idl */,
+				FFEFAB2D183BCC6F00514548 /* SubscriptionObserver.h */,
```
Don't include this change in this PR.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -966,8 +966,23 @@ void NetworkConnectionToWebProcess::registerInternalFileBlobURL(const URL& url,
 
-    m_blobURLs.add({ url, std::nullopt });
-    session->blobRegistry().registerInternalFileBlobURL(url, BlobDataFileReferenceWithSandboxExtension::create(path, replacementPath, SandboxExtension::create(WTFMove(extensionHandle))), contentType);
+    auto after = [this, weakThis = WeakPtr { * this }, url, path, replacementPath, extensionHandle = WTFMove(extensionHandle), contentType] (bool allowed) mutable {
```
Can we call this completionHandler or registerInternalFileBlobURLs or continueRegistration or something?

achristensen07 commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -799,11 +799,34 @@ void NetworkStorageManager::didIncreaseQuota(WebCore::ClientOrigin&& origin, Quo
+    ASSERT(RunLoop::isMain());
+    ASSERT(!m_closed);
+
+    m_queue->dispatch([this, protectedThis = Ref { *this }, connectionID, completionHandler = WTFMove(completionHandler), path = path.isolatedCopy()]() mutable {
```
Why is this on a background queue?  It's just doing string operations, not file I/O, right?

pascoej replied:
`m_temporaryBlobPathsByConnection` may only be accessed on the background queue. 

`WTF_GUARDED_BY_CAPABILITY(workQueue());`

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -579,6 +583,7 @@ class NetworkProcess final : public AuxiliaryProcess, private DownloadManager::C
 
     bool m_privateClickMeasurementEnabled { true };
     bool m_ftpEnabled { false };
+    bool m_blobFileAccessEnforceEnabled { false };
```
Enforcement

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
```diff
@@ -215,6 +215,9 @@ messages -> NetworkProcess LegacyReceiver {
     CountNonDefaultSessionSets(PAL::SessionID sessionID) -> (size_t count)
 
+    AllowFileAccessFromWebProcess(WebCore::ProcessIdentifier processIdentifier, String url) -> ()
+    AllowFilesAccessFromWebProcess(WebCore::ProcessIdentifier processIdentifier, Vector<String> urls) -> ()
```
Can we make it so there is only one message, and just send a vector of 1 string for the single case?

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3181,33 +3210,48 @@ void WebPageProxy::executeEditCommand(const String& commandName, const String& a
+    if (auto pasteAccessCategory = pasteAccessCategoryForCommand(commandName))
+        willPerformPasteCommand(*pasteAccessCategory, WTFMove(after), frameID);
+    else
+        after();
```
Yeah, I'm not sure how I feel about the name "after".  "completionHandler" would be more consistent with the rest of our code.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3181,33 +3210,48 @@ void WebPageProxy::executeEditCommand(const String& commandName, const String& a
 
-    if (auto pasteAccessCategory = pasteAccessCategoryForCommand(commandName))
-        willPerformPasteCommand(*pasteAccessCategory, focusedFrame->frameID());
+    auto after = [weakThis = WeakPtr { *this }, commandName = commandName.isolatedCopy(), argument = argument.isolatedCopy(), frameID] () mutable {
```
Why isolated copy?  Are we changing threads?

achristensen07 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -4792,7 +4792,7 @@ - (BOOL)_handleDOMPasteRequestWithResult:(WebCore::DOMPasteAccessResponse)respon
     if (auto pasteAccessCategory = std::exchange(_domPasteRequestCategory, std::nullopt)) {
         if (response == WebCore::DOMPasteAccessResponse::GrantedForCommand || response == WebCore::DOMPasteAccessResponse::GrantedForGesture)
-            _page->grantAccessToCurrentPasteboardData(pasteboardNameForAccessCategory(*pasteAccessCategory));
+            _page->grantAccessToCurrentPasteboardData(pasteboardNameForAccessCategory(*pasteAccessCategory), [] () { });
```
Do we need the rest of this function to be inside the completion handler?

achristensen07 commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -237,7 +237,7 @@ static inline bool expectsLegacyImplicitRubberBandControl()
         return false;
 
-    grantAccessToCurrentPasteboardData(pasteboardName);
+    grantAccessToCurrentPasteboardData(pasteboardName, [] () { });
```
Do we need the rest of this function inside the completion handler?

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PasteHTML.mm:
```diff
@@ -79,6 +79,7 @@ void writeHTMLToPasteboard(NSString *html)
 
     writeHTMLToPasteboard(@"<!DOCTYPE html><html><body><p><u>hello</u>, world</p></body></html>");
     [webView paste:nil];
+    TestWebKitAPI::Util::runFor(50_ms);
```
This seems like an indication of an issue and looks like it will introduce flaky tests.  Can we instead wait for the operation we are actually waiting for?  Is this a bincompat issue?

achristensen07 commented on Source/WebCore/Modules/indexeddb/server/IDBBackingStore.h:
```diff
@@ -28,6 +28,7 @@
 #include "IDBDatabaseInfo.h"
 #include "IDBError.h"
 #include "IndexKey.h"
+#include "ProcessIdentifier.h"
```
This seems unnecessary.

achristensen07 commented on Source/WebCore/Modules/indexeddb/server/SQLiteIDBTransaction.cpp:
```diff
@@ -83,6 +83,16 @@ IDBError SQLiteIDBTransaction::commit()
+{
+    String databaseDirectory = m_backingStore.databaseDirectory();
+
+    Vector<String> paths = WTF::map(m_blobTemporaryAndStoredFilenames, [databaseDirectory = WTFMove(databaseDirectory)](auto& entry) -> String {
```
nit: return WTF::map(...) would be a little more concise.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1070,6 +1099,25 @@ void NetworkConnectionToWebProcess::writeBlobsToTemporaryFilesForIndexedDB(const
+void NetworkConnectionToWebProcess::allowAccessToFile(const String& path)
+{
+    if (auto* session = networkSession())
+        session->protectedStorageManager()->registerTemporaryBlobFilePaths(Ref { m_connection }, { path });
```
I think Ref { } is unnecessary.  Same in the Vector function below.

achristensen07 commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -799,11 +799,34 @@ void NetworkStorageManager::didIncreaseQuota(WebCore::ClientOrigin&& origin, Quo
+
+        auto& temporaryBlobPathsForConnection = it->value;
+        bool containsPath = temporaryBlobPathsForConnection.contains(path);
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis), containsPath, completionHandler = WTFMove(completionHandler)]() mutable {
```
I'm not sure if we still need protectedThis to call a completion handler that doesn't need this at all.  We don't protect this in the false case above.

achristensen07 commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -237,7 +238,8 @@ static inline bool expectsLegacyImplicitRubberBandControl()
 
-    grantAccessToCurrentPasteboardData(pasteboardName);
+    if (auto replyID = grantAccessToCurrentPasteboardData(pasteboardName, [] () { }))
+        websiteDataStore().protectedNetworkProcess()->connection()->waitForAsyncReplyAndDispatchImmediately<Messages::NetworkProcess::AllowFilesAccessFromWebProcess>(*replyID, 100_ms);
```
I don't think we want to introduce calls to waitForAsyncReplyAndDispatchImmediately, which will introduce hangs.

pascoej replied:
There's already a hang in this function below and the below sync call won't work if this one has not completed.

whsieh replied:
Yeah, I don't think this really makes the problem _worse_ (I do agree this sync IPC isn't ideal, though I'm not sure how much we can do about it, since it's public API we need to continue supporting).

I also imagine that sync IPC between the network process and UI process (both of which should not be compromised by maliciously crafted content) is much less bad than blocking on the untrusted web content process. A short 100 ms timeout seems reasonable.

whsieh replied:
The thing I'd actually be more concerned about is cross-process IPC deadlocks — for instance, the UI process waiting for the network process while the network process simultaneously syncwaits for the UI process. Hopefully this isn't an issue, and the sync IPC only goes in one direction...

pascoej replied:
To close the loop above. I think the only non-test message that the network process synchronously sends the UI process / waits for is `SecItemShimProxy::SecItemRequestSync` which I think would never be called during a `AllowFilesAccessFromWebProcess.`

achristensen07 commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4079,6 +4080,64 @@ static NSDragOperation kit(std::optional<WebCore::DragOperation> dragOperation)
+    networkProcess->sendWithAsyncReply(Messages::NetworkProcess::AllowFilesAccessFromWebProcess(page->protectedLegacyMainFrameProcess()->coreProcessIdentifier(), *fileNames), [page = WTFMove(page), fileNames, dragData, pasteboardName]() mutable {
+        if (!page)
+            return;
+        ensureOnMainRunLoop([page = WTFMove(page), fileNames, dragData, pasteboardName]() mutable {
```
I'm pretty sure this is unnecessary.  sendWithAsyncReply should call the lambda on the main run loop.

pascoej commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4488,7 +4507,7 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
 {
     if (response == WebCore::DOMPasteAccessResponse::GrantedForCommand || response == WebCore::DOMPasteAccessResponse::GrantedForGesture)
-        m_page->grantAccessToCurrentPasteboardData(pasteboardNameForAccessCategory(pasteAccessCategory));
+        m_page->grantAccessToCurrentPasteboardData(pasteboardNameForAccessCategory(pasteAccessCategory), [] () { });
```
This is only used by a context item where it doesn't perform an action after (WKDOMPasteMenuDelegate).

szewai reviewed (changes requested):
`m_temporaryBlobPathsByConnection` is used specifically for IndexedDB temp blob files (the tmp/ file is not expected to be used in other cases with blob url), and the access is expected to be revoked when the temp files are moved and the corresponding request is handled, see `NetworkStorageManager::putOrAdd`. 
I don't think it should be used for general file-backed blobs, especially it's only accessed from `NetworkStorageManager`'s background queue, which is not required by other blob usage.

You may want to add the map to where the check will be performed.

szewai commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageManager.cpp:
```diff
@@ -90,6 +91,9 @@ Expected<WebCore::FileSystemHandleIdentifier, FileSystemStorageError> FileSystem
         }
     }
 
+    if (!fileExists && createIfNecessary)
```
I don't quite understand this change. File System API does not directly work with blob, and this patch does not validate that file paths are accessible in subsequent File System requests. 
(Also subsequent requests only contain identifier, so I don't understand how this can be useful...)

pascoej replied:
At one point I thought I needed this change to make `FileSystemDirectoryHandle.getFileHandle(..., {create = true})` work properly, but I don't think it's needed anymore. I've removed it and made the other changes requested.

pascoej replied:
Actually I think this is still needed as tests like mentioned above (ex. imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any.html ) will not work without this. It wasn't showing up before due to an error piping through the feature flag. I'm going to reincorporate this.

szewai replied:
Checking the test, there is indeed one case where a file-backed blob can be created from path returned File System API: `FileSystemFileHandle::getFile()`. 
We can make network process adds the path to allowlist before replying the message. i.e. Make `NetworkStorageManager::getFile()` dispatch a task to main thread and ask `m_process` to add the path for the connection. Note `void NetworkStorageManager::getFile` should take `IPC::Connection&` as it first parameter.

szewai commented on Source/WebCore/Modules/indexeddb/server/UniqueIDBDatabase.h:
```diff
@@ -112,6 +112,8 @@ class UniqueIDBDatabase : public CanMakeWeakPtr<UniqueIDBDatabase> {
     void commitTransaction(UniqueIDBDatabaseTransaction&, uint64_t handledRequestResultsCount, ErrorCallback&&, SpaceCheckResult = SpaceCheckResult::Unknown);
     void abortTransaction(UniqueIDBDatabaseTransaction&, ErrorCallback&&, SpaceCheckResult = SpaceCheckResult::Unknown);
 
+    Vector<String> storedBlobPathsForTransaction(UniqueIDBDatabaseTransaction&);
```
Where is this used? What's the purpose of this change?

pascoej replied:
It's vestigial. I will remove it.

szewai reviewed (commented):
It's a bit unclear to me what FileSystem and IndexedDB changes are for.

szewai commented on Source/WebCore/Modules/indexeddb/server/IDBBackingStore.h:
```diff
@@ -98,6 +98,8 @@ class IDBBackingStore {
     virtual bool hasTransaction(const IDBResourceIdentifier&) const = 0;
     virtual void handleLowMemoryWarning() = 0;
 
+    virtual Vector<String> storedFilesForTransaction(const IDBResourceIdentifier& transactionIdentifier) = 0;
```
The indexeddb changes should be reverted.

szewai commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageManager.h:
```diff
@@ -41,6 +41,7 @@ namespace WebKit {
 
 class FileSystemStorageHandle;
 class FileSystemStorageHandleRegistry;
+class NetworkStorageManager;
```
The changes in Source/WebKit/NetworkProcess/storage/ should be reverted.

szewai commented on Source/WebCore/platform/DragData.h:
```diff
@@ -108,7 +108,7 @@ class DragData {
     String asURL(FilenameConversionPolicy = ConvertFilenames, String* title = nullptr) const;
     String asPlainText() const;
-    Vector<String> asFilenames() const;
+    WEBCORE_EXPORT Vector<String> asFilenames() const;
```
Is this changed needed?

szewai commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -393,6 +393,8 @@ class ChromeClient {
     virtual RefPtr<ShapeDetection::FaceDetector> createFaceDetector(const ShapeDetection::FaceDetectorOptions&) const { return nullptr; }
     virtual RefPtr<ShapeDetection::TextDetector> createTextDetector() const { return nullptr; }
 
+    virtual void registerBlobPathForTesting(const String&) { }
```
Why making it virtual? Who implements this?

szewai reviewed (commented):
I think we probably need some cleanup to keep only the necessary changes.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -286,6 +290,7 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver, public
     UnifiedOriginStorageLevel m_unifiedOriginStorageLevel;
     IPC::Connection::UniqueID m_parentConnection;
     HashMap<IPC::Connection::UniqueID, HashSet<String>> m_temporaryBlobPathsByConnection WTF_GUARDED_BY_CAPABILITY(workQueue());
+    HashMap<IPC::Connection::UniqueID, HashSet<String>> m_allowedFilePathsByConnection WTF_GUARDED_BY_CAPABILITY(workQueue());
```
`m_allowedFilePathsByConnection` should be accessed on main thread, not work queue because there is only one case where the background queue operation will change it: `NetworkStorageManager::getFile()`; all the other cases (`NetworkConnectionToWebProcess::registerInternalFileBlobURL()`, `NetworkProcess::allowFileAccessFromWebProcess()`, etc) are on main thread.

Also, I don't think we need to add the path when File System handle is created, since only `FileSystemFileHandle::getFile()` will create file backed blob. That means we only need to make `NetworkStorageManager::getFile()` dispatch a task to main thread to add the path to `m_allowedFilePathsByConnection`.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -136,6 +136,7 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver, public
     void clearServiceWorkerRegistrations(CompletionHandler<void()>&&);
     void importServiceWorkerRegistrations(CompletionHandler<void(std::optional<Vector<WebCore::ServiceWorkerContextData>>)>&&);
     void updateServiceWorkerRegistrations(Vector<WebCore::ServiceWorkerContextData>&&, Vector<WebCore::ServiceWorkerRegistrationKey>&&, CompletionHandler<void(std::optional<Vector<WebCore::ServiceWorkerScripts>>)>&&);
+    const String& path() const { return m_path; }
```
Note `m_path` is currently set and accessed on WorkQueue, so we should either protect it with lock, or introduce a `m_mainThreadPath` which is only accessed from main thread (set it in constructor) and return `m_mainThreadPath` here.

fujii commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5366,8 +5367,37 @@ RefPtr<File> Internals::createFile(const String& path)
+    }
+
+    if (auto* page = document->page()) {
+        page->chrome().client().registerBlobPathForTesting(url.fileSystemPath(), [promise = WTFMove(promise), weakDocument = WeakPtr { *document }, url = WTFMove(url)] () mutable {
```
This `url` is use-after-move on Windows port. We need `auto path = url.fileSystemPath();` before `WTFMove(url)` line.

pascoej committed:
Add validation to file-backed blob urls from WebContent
https://bugs.webkit.org/show_bug.cgi?id=276193
rdar://80200949

Reviewed by Alex Christensen.

This validates that file paths are valid for a given WebContent process when registering the blob

* LayoutTests/editing/pasteboard/copy-paste-attachment.html:
* LayoutTests/fast/attachment/attachment-dom.html:
* LayoutTests/fast/attachment/attachment-folder-icon-expected.html:
* LayoutTests/fast/attachment/attachment-icon-from-file-extension-expected.html:
* LayoutTests/fast/attachment/attachment-label-highlight.html:
* LayoutTests/fast/attachment/attachment-progress.html:
* LayoutTests/fast/attachment/attachment-select-on-click-inside-user-select-all.html:
* LayoutTests/fast/attachment/attachment-select-on-click.html:
* LayoutTests/fast/attachment/attachment-subtitle.html:
* LayoutTests/fast/attachment/attachment-title.html:
* LayoutTests/fast/attachment/attachment-type-attribute-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-dom.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension-expected.html:
* LayoutTests/fast/attachment/mac/attachment-keynote-icon-expected.html:
* LayoutTests/fast/attachment/mac/attachment-numbers-icon-expected.html:
* LayoutTests/fast/attachment/mac/attachment-pages-icon-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute-expected.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-convert-to-download-async-delegate.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-convert-to-download.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-blank-base-target-popup-not-allowed.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-blank-target-popup-not-allowed.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-blank-target.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-backslash.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-doublequote.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-slashes.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-includes-unicode.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download-no-extension.html:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-file-blob-download.html:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::registerBlobPathForTesting):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::createFile):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::isFilePathAllowed):
(WebKit::NetworkConnectionToWebProcess::registerInternalFileBlobURL):
(WebKit::NetworkConnectionToWebProcess::registerInternalBlobURLOptionallyFileBacked):
(WebKit::NetworkConnectionToWebProcess::registerBlobPathForTesting):
(WebKit::NetworkConnectionToWebProcess::allowAccessToFile):
(WebKit::NetworkConnectionToWebProcess::allowAccessToFiles):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
(WebKit::NetworkConnectionToWebProcess::blobFileAccessEnforcementEnabled const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::allowFilesAccessFromWebProcess):
(WebKit::NetworkProcess::allowFileAccessFromWebProcess):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::grantAccessToCurrentPasteboardData):
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::grantAccessToCurrentData):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::maybeInitializeSandboxExtensionHandle):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadDataWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::reload):
(WebKit::WebPageProxy::executeEditCommand):
(WebKit::WebPageProxy::performDragOperation):
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::contextMenuItemSelected):
(WebKit::WebPageProxy::didChooseFilesForOpenPanel):
(WebKit::WebPageProxy::platformDidSelectItemFromActiveContextMenu):
(WebKit::WebPageProxy::willPerformPasteCommand):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURL):
(WebKit::WebProcessProxy::assumeReadAccessToBaseURLs):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _handleDOMPasteRequestWithResult:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::willPerformPasteCommand):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::readSelectionFromPasteboard):
(WebKit::WebPageProxy::platformDidSelectItemFromActiveContextMenu):
(WebKit::WebPageProxy::willPerformPasteCommand):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::performDragWithLegacyFiles):
(WebKit::handleLegacyFilesPasteboard):
(WebKit::WebViewImpl::performDragOperation):
(WebKit::WebViewImpl::requestDOMPasteAccess):
(WebKit::WebViewImpl::handleDOMPasteRequestForCategoryWithResult):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::registerBlobPathForTesting):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::registerBlobPathForTesting):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(TestWebKitAPI::TEST(WebKit, RunOpenPanelNonLatin1)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UploadDirectory.mm:
(TEST(WebKit, UploadDirectory)):

Canonical link: https://commits.webkit.org/281966@main

carlosgcampos commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -958,13 +958,28 @@ void NetworkConnectionToWebProcess::cookieEnabledStateMayHaveChanged()
 
+bool NetworkConnectionToWebProcess::isFilePathAllowed(NetworkSession& session, String path)
+{
+    path = FileSystem::lexicallyNormal(path);
+    auto parentPath = FileSystem::parentPath(path);
+    while (parentPath != path) {
+        if (m_allowedFilePaths.contains(path) || parentPath == session.storageManager().path() || parentPath == session.storageManager().customIDBStoragePath())
+            return true;
+        path = parentPath;
+        parentPath = FileSystem::parentPath(path);
+    }
+    return false;
```
I don't think this always works. When the path was allowed from WebProcessProxy::assumeReadAccessToBaseURL() the base path contains a trailing slash, but here FileSystem::parentPath() returns the parent directory without a trailing slash. Since we compare strings here, `/tmp/ != /tmp`
